vm/qemu/qemu.go

```diff
@@ -671,6 +673,10 @@ func (inst *instance) Diagnose() ([]byte, bool) {
 					inst.index, err)
 		}
 	}
+	// The caller is going to kill the process anyway but if
+	// QEMU was running from a script, Process.Kill() will kill
+	// the script and not QEMU
+	inst.qmp(`{"execute":"quit"}`)
```

Comment:
The commit description says this will wait for the kernel output to finish, but how does this work? How does qemu know when kernel finishes the debug output?

Kernel halts after finishing the panic output. So if qemu exits on halt, I can imagine how it all works. But now we don't shutdown and instead asynchronously execute quit. So my naive thinking would be that now we truncate output.
Please also document in a comment why we are doing -no-shutdown and how this exiting sequence works now.

Reply:
```
-	return nil, false
+	return ret, true
``` this chunk was about waiting.
Calling "quit" has nothing to do with it, the comment above says what it does :)

Reply:
Reading `extractError` in `vm/vm.go` it seems that returning true here only makes sense if Diagnose triggers any asynchronous output on the kernel console (then extractError will wait for it). Use of the debug protocol does not trigger any output on the kernel console, so returning true will only unnecessary delay execution.

And I am still puzzled from this patch is doing, why using the exit command is better and why we need to use -no-shutdown.

Reply:
I remember now why I added it - I was chasing one bug and I had "ftrace=irqsoff ftrace_dump_on_oops" which dumps a lot on panic. Not needed in a general case.

The reason for explicit "quit" is described in the comment. I run QEMU from a script which allows me adding more stuff which cannot be added directly to the syz-manager's config - one example is a gdb server in QEMU which requires a free tcp port (as we run many instances of QEMU in parallel) so the script picks one. So when it is a script, and syzkaller kills a VM - it kills a script but not QEMU process so they remain in memory. Explicit "quit" kills the QEMU process. Alternative is killing all the children of the initial script.

Reply:
Using the exit command instead of kill for this somewhat hacky. It will never be tested and eventually we want to unify more code between VM impls, and it will be useful to assume that a vm starts a process that can be killed.

Is it possible to a shell script to enforce PR_SET_PDEATHSIG on child processes? Maybe there is some similar flag that a parent (shell script) can set of itself?

Or will it work if we start qemu in a new process group and then kill -PID (should kill all processes in the process group).

Or perhaps it's better to improve vm/qemu configuration flexibility to support what you are doing the shell script. See e.g. what we do for `{{TEMPLATE}}` in vm/qemu, we could add something like `{{TCP_PORT}}`.

Reply:
The "exit" command does not replace killing the task, it is just an additional step which should not affect the normal way of syzkaller operation.

Using process group will do too, I just could not figure out an easy/quick way to do so, it looked to me as there is support but something did not work as expected.

I'll look at PR_SET_PDEATHSIG and {{TEMPLATE}}, {{TCP_PORT}} should be useful regardless, and most likely I'll drop the quitting patch for now.

Reply:
If {{TCP_PORT}} will work, that looks like a useful feature anyway.

But for process group, you used Setpgid when starting the qemu process:
https://golang.org/pkg/syscall/#SysProcAttr
and did kill -pid, right?